### PR TITLE
[FEATURE] Add minify task and processor

### PR DESCRIPTION
**Pull request by RandomByte**
Implements https://github.com/SAP/ui5-tooling/pull/583 based on https://github.com/SAP/ui5-builder/pull/657

**Depends on https://github.com/SAP/ui5-fs/pull/331**

- Combine debug-file creation, minification and source-map generation into a single `minify` task
    - Replaces existing `uglify` and `createDebugFiles` tasks
- Tag debug-files with new "global" (=shared across projects) tag `IsDebugVariant`
- Filter debug-files from moduleBundler so that it only packages minified or debug-sources (depending on `optimize` bundleOption)

### Open TODOs:
* [x] Add integration test (or modify an existing one) for `optimize: true` bundle generation **without** minify task
* [x] Add integration test (or modify an existing one) for `optimize: false` bundle generation **with** minify task

### CODE REVIEW

File: test/lib/index.js [hunk 1]
```diff
@@ -28,7 +26,7 @@ test("index.js exports all expected modules", (t) => {
 	t.truthy(index.tasks.generateBundle, "Module exported");
 	t.truthy(index.tasks.generateCachebusterInfo, "Module exported");
 	t.truthy(index.tasks.buildThemes, "Module exported");
-	t.truthy(index.tasks.createDebugFiles, "Module exported");
+	t.truthy(index.tasks.minifier, "Module exported");
```
Comment (matz3) [inline test/lib/index.js:29]:
```suggestion
	t.truthy(index.tasks.minify, "Module exported");
```

**Review by matz3 — CHANGES_REQUESTED** [review]

File: test/lib/index.js (hunk 1 above)
Comment (RandomByte) [reply test/lib/index.js]: Done

File: lib/lbt/resources/LocatorResourcePool.js [hunk 2]
```diff
@@ -2,6 +2,9 @@ const ResourcePool = require("./ResourcePool");
 const LocatorResource = require("./LocatorResource");
 
 class LocatorResourcePool extends ResourcePool {
+	constructor({ignoreMissingModules} = {}) {
```
Comment (matz3) [inline lib/lbt/resources/LocatorResourcePool.js:5]: Unnecessary change, right?

File: lib/lbt/resources/ResourcePool.js [hunk 3]
```diff
@@ -58,7 +59,7 @@ function scanFileOrDir(fileOrDir, name, pool) {
 */
 
 async function determineDependencyInfo(resource, rawInfo, pool) {
-	const info = new ModuleInfo(resource.name);
+	const info = new ModuleInfo(ModuleName.getNonDebugName(resource.name) || resource.name);
```
Comment (matz3) [inline lib/lbt/resources/ResourcePool.js:62]: Do we still need this?

File: lib/processors/minifier.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+ * @returns {Promise<object>} Promise resolving with object of resource, dbgResource and sourceMap
+ */
+module.exports = async function({resources}) {
+	return Promise.all(resources.map(async (resource) => {
+		const dbgPath = resource.getPath().replace(debugFileRegex, "-dbg$1");
+		const dbgResource = await resource.clone();
+		dbgResource.setPath(dbgPath);
+
+		const filename = path.posix.basename(resource.getPath());
+		if (filename.endsWith(".support.js")) {
```
Comment (matz3) [inline lib/processors/minifier.js:34]: AFAIK this is the reason why it should not be done
```suggestion
		// Support rules should not be minified to have readable code in the Support Assistant
		if (filename.endsWith(".support.js")) {
```

File: lib/tasks/minify.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * @param {string} parameters.options.pattern Pattern to locate the files to be processed
+ * @returns {Promise<undefined>} Promise resolving with <code>undefined</code> once data has been written
+ */
+module.exports = async function({workspace, taskUtil, options: {pattern}}) {
+	const resources = await workspace.byGlob(pattern);
+	const processedResources = await minifier({resources});
+
+	return Promise.all(processedResources.map(async ({resource, dbgResource, sourceMapResource}) => {
+		if (taskUtil) {
+			taskUtil.setTag(resource, taskUtil.STANDARD_TAGS.HasDebugVariant);
```
Comment (matz3) [inline lib/tasks/minify.js:21]: Is this really correct for the "*.support.js" files?
I would expect that those should be skipped completely (no tags set).

File: test/lib/processors/minifier.js [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+	t.deepEqual(await resources[2].dbgResource.getString(), content3, "Correct debug content for resource 3");
+	t.deepEqual(await resources[2].sourceMapResource.getPath(), "/test3.designtime.js.map",
+		"Correct resource path for source map content of resource 3");
+	t.deepEqual(await resources[2].sourceMapResource.getString(), expectedSourceMap3,
+		"Correct source map content for resource 3");
+
+	t.deepEqual(await resources[3].resource.getPath(), "/test4.support.js",
+		"Correct resource path for minified content of resource 4");
+	t.deepEqual(await resources[3].resource.getString(), expectedMinified4, "Correct minified content for resource 4");
+	t.deepEqual(await resources[3].dbgResource.getPath(), "/test4-dbg.support.js",
```
Comment (matz3) [inline test/lib/processors/minifier.js]: This assertion should fail, right?

File: test/lib/processors/minifier.js [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+		`{"version":3,"sources":["test1-dbg.controller.js"],"names":["test1","paramA","variableA","console","log"],` +
+		`"mappings":"AACA,SAASA,MAAMC,GACd,IAAIC,EAAYD,EAChBE,QAAQC,IAAIF,GAEbF","file":"test1.controller.js"}`;
+	const expectedSourceMap2 =
+		`{"version":3,"sources":["test2-dbg.fragment.js"],"names":["test2","paramA","variableA","console","log"],` +
+		`"mappings":"AACA,SAASA,MAAMC,GACd,IAAIC,EAAYD,EAChBE,QAAQC,IAAIF,GAEbF","file":"test2.fragment.js"}`;
+	const expectedSourceMap3 =
+		`{"version":3,"sources":["test3-dbg.designtime.js"],"names":["test3","paramA","variableA","console","log"],` +
+		`"mappings":"AACA,SAASA,MAAMC,GACd,IAAIC,EAAYD,EAChBE,QAAQC,IAAIF,GAEbF","file":"test3.designtime.js"}`;
+
+	t.deepEqual(await resources[0].resource.getPath(), "/test1.controller.js",
```
Comment (matz3) [inline test/lib/processors/minifier.js]: `getPath` isn't async, right?

File: lib/tasks/bundlers/generateLibraryPreload.js [hunk 8]
```diff
@@ -296,6 +304,27 @@ module.exports = function({workspace, dependencies, taskUtil, options: {projectN
...
+				if (taskUtil) {
+					unoptimizedResources = await new ReaderCollectionPrioritized({
+						name: `libraryBundler - prioritize workspace over dependencies: ${projectName}`,
+						readers: [workspace, dependencies]
+					}).filter(function(resource) {
+						// Remove any non-debug variants
+						return !taskUtil.getTag(resource, taskUtil.STANDARD_TAGS.HasDebugVariant);
+					}).transformer(async function(resourcePath, getClonedResource) {
+						if (taskUtil.getTag(resourcePath, taskUtil.STANDARD_TAGS.IsDebugVariant)) {
+							const resource = await getClonedResource();
```
Comment (matz3) [inline lib/tasks/bundlers/generateLibraryPreload.js:318]:
```suggestion
					}).transformer(async function(resourcePath, getResource) {
						if (taskUtil.getTag(resourcePath, taskUtil.STANDARD_TAGS.IsDebugVariant)) {
							const resource = await getResource();
```

**Comment by RandomByte** [comment]
Tests are failing because of dependency to https://github.com/SAP/ui5-fs/pull/331

File: test/expected/build/application.i/dest/Component-preload.js [hunk 9]
```diff
@@ -3,9 +3,11 @@ jQuery.sap.registerPreloadedModules({
 "version":"2.0",
 "modules":{
 	"application/i/Component.js":function(){sap.ui.define(["sap/ui/core/UIComponent"],function(n){"use strict";return n.extend("application.i.Component",{metadata:{manifest:"json"}})});
+//# sourceMappingURL=Component.js.map
```
Comment (matz3) [inline test/expected/build/application.i/dest/Component-preload.js:6]: Should those comments be part of the preload? If the browser ignores them I guess it's okay. If it causes problems, we should remove them.

File: test/expected/build/library.ø/dest/resources/library/ø/rules/MyControl-dbg.support.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
+/*!
```
Comment (matz3) [inline test/expected/build/library.ø/dest/resources/library/ø/rules/MyControl-dbg.support.js]: As already mentioned above: We don't want to create -dbg files for *.support.js files. So they should just remain as-is.

**Review by matz3 — CHANGES_REQUESTED** [review]

File: lib/lbt/resources/LocatorResourcePool.js (hunk 2 above)
Comment (RandomByte) [reply lib/lbt/resources/LocatorResourcePool.js]: Done

File: lib/lbt/resources/ResourcePool.js (hunk 3 above)
Comment (RandomByte) [reply lib/lbt/resources/ResourcePool.js]: Done

File: lib/processors/minifier.js (hunk 4 above)
Comment (RandomByte) [reply lib/processors/minifier.js]: Done, moved to glob pattern in Application- and LibraryBuilder

File: lib/tasks/minify.js (hunk 5 above)
Comment (RandomByte) [reply lib/tasks/minify.js:21]: Done

File: test/lib/processors/minifier.js (hunk 6 above)
Comment (RandomByte) [reply test/lib/processors/minifier.js]: Not relevant anymore, minifier now has no special handling of support files.

File: test/lib/processors/minifier.js (hunk 7 above)
Comment (RandomByte) [reply test/lib/processors/minifier.js]: Done

File: lib/tasks/bundlers/generateLibraryPreload.js (hunk 8 above)
Comment (RandomByte) [reply lib/tasks/bundlers/generateLibraryPreload.js]: Done

File: test/expected/build/application.i/dest/Component-preload.js (hunk 9 above)
Comment (RandomByte) [reply test/expected/build/application.i/dest/Component-preload.js:6]: I *think* they are not a problem unless they are at the end of the file. Anyways, the source map change for the module bundler will remove them 👍

File: test/expected/build/library.ø/dest/resources/library/ø/rules/MyControl-dbg.support.js (hunk 10 above)
Comment (RandomByte) [reply test/expected/build/library.ø/dest/resources/library/ø/rules/MyControl-dbg.support.js]: Done

**Comment by RandomByte** [comment]
Rebased on top of latest `next` to get new ui5-fs features

File: lib/tasks/TaskUtil.js [hunk 11]
```diff
@@ -68,14 +68,14 @@ class TaskUtil {
...
+	 * @param {string|module:@ui5/fs.Resource} resourcePath Path or resource-instance the tag should be retrieved for
 	 * @param {string} tag Name of the tag
 	 * @returns {string|boolean|integer|undefined} Tag value for the given resource.
 	 * 										<code>undefined</code> if no value is available
 	 * @public
 	 */
-	getTag(resource, tag) {
-		return this._projectBuildContext.getResourceTagCollection().getTag(resource, tag);
+	getTag(resourcePath, tag) {
+		return this._projectBuildContext.getResourceTagCollection().getTag(resourcePath, tag);
```
Comment (matz3) [inline lib/tasks/TaskUtil.js:78]:
```suggestion
		return this.getResourceTagCollection().getTag(resourcePath, tag);
```

File: lib/tasks/TaskUtil.js [hunk 12]
```diff
@@ -86,12 +86,16 @@ class TaskUtil {
...
 	 *
-	 * @param {module:@ui5/fs.Resource} resource The resource the tag should be cleared for
+	 * @param {string|module:@ui5/fs.Resource} resourcePath Path or resource-instance the tag should be cleared for
 	 * @param {string} tag Tag
 	 * @public
 	 */
-	clearTag(resource, tag) {
-		return this._projectBuildContext.getResourceTagCollection().clearTag(resource, tag);
+	clearTag(resourcePath, tag) {
+		return this._projectBuildContext.getResourceTagCollection().clearTag(resourcePath, tag);
```
Comment (matz3) [inline lib/tasks/TaskUtil.js:94]:
```suggestion
		return this.getResourceTagCollection().clearTag(resourcePath, tag);
```

File: lib/tasks/TaskUtil.js [hunk 13]
```diff
@@ -86,12 +86,16 @@ class TaskUtil {
...
+	}
+
+	getResourceTagCollection() {
```
Comment (matz3) [inline lib/tasks/TaskUtil.js:97]: But.. is this method actually used somewhere?
I couldn't find a reference.

File: lib/tasks/bundlers/generateStandaloneAppBundle.js [hunk 14]
```diff
@@ -76,12 +77,18 @@ module.exports = async function({workspace, dependencies, taskUtil, options: {pr
 			`unable to generate complete bundles for such projects.`);
 	}
 
-	// If an application does not have a namespace, its resources are located at the root. Otherwise in /resources
```
Comment (matz3) [inline lib/tasks/bundlers/generateStandaloneAppBundle.js:79]: By removing this behavior, we need to make sure to fail in case we don't have a namespace (as already planned for 3.0)

File: lib/tasks/TaskUtil.js (hunk 13 above)
Comment (RandomByte) [reply lib/tasks/TaskUtil.js:97]: Right, this shouldn't be necessary anymore. Probably it was added for some different, intermediate implementation.

File: lib/tasks/TaskUtil.js [hunk 15]
```diff
@@ -51,14 +51,14 @@ class TaskUtil {
...
+	 * @param {string|module:@ui5/fs.Resource} resourcePath Path or resource-instance the tag should be stored for
 	 * @param {string} tag Name of the tag.
 	 * 		Currently only the [STANDARD_TAGS]{@link module:@ui5/builder.tasks.TaskUtil#STANDARD_TAGS} are allowed
 	 * @param {string|boolean|integer} [value=true] Tag value. Must be primitive
 	 * @public
 	 */
-	setTag(resource, tag, value) {
-		return this._projectBuildContext.getResourceTagCollection().setTag(resource, tag, value);
+	setTag(resourcePath, tag, value) {
+		return this._projectBuildContext.getResourceTagCollection().setTag(resourcePath, tag, value);
```
[the remark this answers is not in the document]
Comment (RandomByte) [reply lib/tasks/TaskUtil.js:61]: Removed the (unused) getResourceTagCollection function

File: lib/tasks/TaskUtil.js (hunk 12 above)
Comment (RandomByte) [reply lib/tasks/TaskUtil.js:94]: Removed the (unused) getResourceTagCollection function

File: lib/tasks/bundlers/generateStandaloneAppBundle.js (hunk 14 above)
Comment (RandomByte) [reply lib/tasks/bundlers/generateStandaloneAppBundle.js:79]: Yes, correct!

File: lib/tasks/TaskUtil.js (hunk 11 above)
Comment (RandomByte) [reply lib/tasks/TaskUtil.js:78]: Removed the (unused) getResourceTagCollection function

**Review by matz3 — APPROVED** [review]
